feat(cli): add /chat file commands for session management (#3190)
QwenLM/qwen-code: .qwen/commands/chat.md

---
description: Chat session manager. /chat [-s|-l|-r|-d|-h] [name]
---

# CRITICAL: First check {{args}}, then route

## Step 0: Immediate Validation (MUST execute FIRST)

**Check `{{args}}` right now, before doing anything else:**

1. Is `{{args}}` empty? → **Show Help immediately, STOP**
2. Is `{{args}}` only whitespace? → **Show Help immediately, STOP**
3. Does the first token look like a valid flag? (`-s`, `--save`, `-l`, `--list`, `-r`, `--resume`, `-d`, `--delete`, `-h`, `--help`)
   - **NO** → **Show Help immediately, STOP**
   - **YES** → Continue to Step 1

**⚠️ DO NOT skip this step. DO NOT proceed with any action until you verify `{{args}}`.**

---

## Step 1: Detect Environment

### Language

Read `~/.qwen/settings.json` (Windows: `%USERPROFILE%\.qwen\settings.json`).
Look for `general.language`. Respond in that language. If not found, match the language the user used in their prompt.

### OS Detection

> 

Reviewer: **[Suggestion]** OS detection runs via shell subprocess on every `/chat` invocation, including `/chat -l` (list) and `/chat -h` (help) where it's completely unnecessary. Move OS detection into `chat-resume.md` only, where it's actually needed.

_— qwen3.6-plus via Qwen Code /review_

Run `echo %OS%` (Windows) or `echo $OSTYPE` (Linux/macOS).

- `Windows_NT` → Windows

> - `linux-*` → Linux

Reviewer: **[Critical]** OS detection is unreliable on many shells. `echo %OS%` only works in CMD.exe (not PowerShell, the Windows default). `$OSTYPE` is bash/zsh-specific and will be empty in fish, PowerShell, nushell, etc. This means resume will generate wrong commands or fail entirely on many common shells.

```suggestion
Run `node -e "console.log(process.platform)"` for portable detection across all shells.

- `win32` → Windows
- `linux` → Linux
- `darwin` → macOS
```

_— qwen3.6-plus via Qwen Code /review_

- `darwin*` → macOS

---

## Step 2: Parse and Route

Split `{{args}}` by whitespace. First token = flag. Remaining = name.

| Flag              | Action                                    |
| ----------------- | ----------------------------------------- |
| `-s` / `--save`   | Go to Step 3                              |
| `-l` / `--list`   | Read `chat-list.md` and execute its logic |
| `-r` / `--resume` | Go to Step 3                              |
| `-d` / `--delete` | Go to Step 3                              |
| `-h` / `--help`   | **Show Help immediately, STOP**           |

### Step 3: Validate name (for `-s`, `-r`, `-d`)

Extract the name (everything after the flag).

> - Is name missing, empty, or whitespace only? → **Show Help immediately, STOP**

Reviewer: **[Critical]** Production routing table maps `-s`, `-r`, `-d` to "Go to Step 3" but after validation there's no explicit flag-to-file mapping. The source version has `| -s | chat-save.md |` etc. The production version requires the LLM to infer which file to read from filenames alone, risking it reads the wrong sub-command.

```suggestion
Add after the existing table:

| Flag | Sub-command file |
| -s / --save | chat-save.md |
| -r / --resume | chat-resume.md |
| -d / --delete | chat-delete.md |
```

_— qwen3.6-plus via Qwen Code /review_

    - **NO, not reserved** → Read corresponding sub-command file and execute

---

## Common Rules

| Rule                  | Value                                                                                                   |
| --------------------- | ------------------------------------------------------------------------------------------------------- |
| **Valid name regex**  | `^[a-zA-Z0-9_.-]+$`                                                                                     |
| **Max length**        | 128 characters                                                                                          |
| **Reserved names**    | `.`, `..`, `__proto__`, `constructor`, `prototype`                                                      |
| **Index path**        | `.qwen/chat-index.json` (project root)                                                                  |
| **Index format**      | `{"name": "sessionId", ...}`                                                                            |
| **Session ID source** | Filename (no extension) of `.jsonl` in `~/.qwen/projects/<hash>/chats/`                                 |

> | **Hash calculation**  | Full cwd path, replace `\` and `/` with `-`, lowercase. E.g., `D:\code\qwen-code` → `d--code-qwen-code` |

Reviewer: **[Critical]** Hash calculation documentation is factually incorrect. The docs say "replace `\` and `/` with `-`, lowercase" but the actual `sanitizeCwd()` replaces ALL non-alphanumeric characters with `-`, and lowercasing only happens on Windows (not Linux/macOS). E.g., `/home/user/my.project` → `-home-user-my-project` on Linux. The LLM will compute the wrong project directory path for any path with dots, spaces, or other special characters, breaking save/resume.

```suggestion
| **Hash calculation**  | Full cwd path, replace ALL non-alphanumeric characters with `-`. On Windows only, also convert to lowercase. E.g., `/home/user/my.project` → `-home-user-my-project` (Linux) |
```

_— qwen3.6-plus via Qwen Code /review_

---

## Common Rules

| Rule                  | Value                                                                                                                                                                                                |
| --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Valid name regex**  | `^[a-zA-Z0-9_.-]+$`                                                                                                                                                                                  |
| **Max length**        | 128 characters                                                                                                                                                                                       |
| **Reserved names**    | `.`, `..`, `__proto__`, `constructor`, `prototype`                                                                                                                                                   |

> | **Index path**        | `.qwen/chat-index.json` (project root)                                                                                                                                                               |

Reviewer: **[Critical]** The `<hash>` description here does not match the actual session storage implementation. Qwen Code stores project sessions under `~/.qwen/projects/<sha256(projectRoot)>/chats/` via `getProjectHash()`, not under a path-derived sanitized directory name. As written, `/chat -s` and `/chat -r` can look in the wrong directory and fail to find real sessions.

```suggestion
| **Hash calculation**  | SHA-256 of the full project root path. On Windows only, normalize the path to lowercase before hashing. Session files live under `~/.qwen/projects/<sha256>/chats/`. |
```

_— gpt-5.4 via Qwen Code /review_

### OS Detection (ONLY for `-r`/`--resume`)

**Skip this step for other flags.** Only run when `-r` is detected.

Run `node -e "console.log(process.platform)"`. Works across all shells.

- `win32` → Windows
- `linux` → Linux
- `darwin` → macOS

---

## Step 2: Parse and Route

Split `{{args}}` by whitespace. First token = flag. Remaining = name.

| Flag              | Action                                    | Sub-Command File |
| ----------------- | ----------------------------------------- | ---------------- |
| `-s` / `--save`   | Go to Step 3                              | `chat-save.md`   |
| `-l` / `--list`   | Read `chat-list.md` and execute its logic | `chat-list.md`   |
| `-r` / `--resume` | Go to Step 3                              | `chat-resume.md` |
| `-d` / `--delete` | Check for `-y`/`--force`, then route      | `chat-delete.md` |
| `-h` / `--help`   | **Show Help immediately, STOP**           | —                |

### Step 3: Validate name (for `-s`, `-r`, `-d`)

Extract the name (everything after the flag). Also check for `-y` or `--force` after the name.

> - Is name missing, empty, or whitespace only? → **Show Help immediately, STOP**

Reviewer: **[Critical]** The router extracts the name as everything after the flag before handling `-y` / `--force`. For the documented `/chat -d my-session -y` form, the extracted name becomes `my-session -y`, so validation fails before `chat-delete.md` can skip confirmation.

Parse delete arguments by separating force flags from the positional session name before validation: treat the first non-force token after `-d` as the name, accept `-y` / `--force` as flags, and reject any extra non-force tokens.

_— gpt-5.5 via Qwen Code /review_

| --------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Valid name regex**  | `^[a-zA-Z0-9_.-]+$`                                                                                                                                                  |
| **Max length**        | 128 characters                                                                                                                                                       |
| **Reserved names**    | `.`, `..`, `__proto__`, `constructor`, `prototype`                                                                                                                   |
| **Index path**        | `.qwen/chat-index.json` (project root)                                                                                                                               |
| **Index format**      | `{"name": "sessionId", ...}`                                                                                                                                         |

> | **Session ID source** | Filename (no extension) of `.jsonl` in `~/.qwen/projects/<hash>/chats/`                                                                                              |

Reviewer: **[Critical]** The command specs still point session lookup at `~/.qwen/projects/<hash>/chats/` and define `<hash>` as SHA-256 of the project root. Current chat recordings are stored under `Storage.getProjectDir()/chats`, and `Storage.getProjectDir()` uses `sanitizeCwd(projectRoot)`, not `getProjectHash()`.

As written, `/chat -s` can fail to find the current session and `/chat -r` can reject valid saved sessions because both commands inspect the wrong directory. Please update the command specs/source/tests/docs to use the actual `~/.qwen/projects/<sanitizeCwd(projectRoot)>/chats/` convention, with `sanitizeCwd` lowercasing on Windows and replacing all non-alphanumeric characters with `-`.

_— gpt-5.5 via Qwen Code /review_

---
description: Chat session manager. /chat [-s|-l|-r|-d|-h] [name] [-y|--force]
---

# CRITICAL: First check {{args}}, then route

## Step 0: Immediate Validation (MUST execute FIRST)

**Check `{{args}}` right now, before doing anything else:**

1. Is `{{args}}` empty? → **Show Help immediately, STOP**
2. Is `{{args}}` only whitespace? → **Show Help immediately, STOP**
3. Does the first token look like a valid flag? (`-s`, `--save`, `-l`, `--list`, `-r`, `--resume`, `-d`, `--delete`, `-h`, `--help`)
   - **NO** → invalid flag/unrecognized → **Show Help immediately, STOP**
   - **YES** → Continue to Step 1

**⚠️ DO NOT skip this step. DO NOT proceed with any action until you verify `{{args}}`.**

---

## Step 1: Detect Environment

### Language

> 

Reviewer: **[Critical]** Reading the full user-level `~/.qwen/settings.json` just to determine `general.language` can expose unrelated sensitive configuration to the model context. That file may contain provider credentials, MCP server headers/tokens, environment values, or other auth-related settings. Please avoid loading the full settings file from this Markdown command; use the current conversation language or a safe mechanism that exposes only the language setting.

_— gpt-5.5 via Qwen Code /review_

- `darwin` → macOS

---

## Step 2: Parse and Route

Split `{{args}}` by whitespace. First token = flag. Remaining = raw_args.

| Flag              | Action                                    | Sub-Command File |
| ----------------- | ----------------------------------------- | ---------------- |
| `-s` / `--save`   | Go to Step 3                              | `chat-save.md`   |
| `-l` / `--list`   | Read `chat-list.md` and execute its logic | `chat-list.md`   |
| `-r` / `--resume` | Go to Step 3                              | `chat-resume.md` |
| `-d` / `--delete` | Go to Step 3                              | `chat-delete.md` |
| `-h` / `--help`   | **Show Help immediately, STOP**           | —                |

### Step 3: Validate name (for `-s`, `-r`, `-d`)

**For delete (`-d`):**

1. Parse raw_args to extract name: Filter out `-y` and `--force` flags first, the first remaining token is the name.
2. If name is missing, empty, or whitespace only → **Show Help immediately, STOP**
3. If extra non-flag tokens remain after the first name → **Show Help immediately, STOP**
4. If `-y` or `--force` was found → Set `forceDelete = true`

**For save/resume (`-s`, `-r`):**

> 1. Name = first token in raw_args

Reviewer: **[Suggestion]** Save/resume only use the first token as the name and do not reject extra trailing tokens, unlike delete. For example, `/chat -s good-name unexpected` would operate on `good-name` while silently ignoring the typo. Please apply the same extra-token validation used for delete to save/resume so malformed invocations stop instead of acting on the wrong name.

_— gpt-5.5 via Qwen Code /review_

| --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Valid name regex**  | `^[a-zA-Z0-9_.-]+$`                                                                                                                                       |
| **Max length**        | 128 characters                                                                                                                                            |
| **Reserved names**    | `.`, `..`, `__proto__`, `constructor`, `prototype`                                                                                                        |
| **Index path**        | `.qwen/chat-index.json` (project root)                                                                                                                    |
| **Index format**      | `{"name": "sessionId", ...}`                                                                                                                              |

> | **Session ID source** | Filename (no extension) of `.jsonl` in `~/.qwen/projects/<sanitizeCwd>/chats/`                                                                            |

Reviewer: **[Critical]** This session path assumes the runtime base is always `~/.qwen`, but Qwen Code stores runtime data under `Storage.getRuntimeBaseDir()`, which can be changed by `QWEN_RUNTIME_DIR` or `advanced.runtimeOutputDir`. Users with either setting configured will have sessions under a different base directory, so `/chat -s` and `/chat -r` can look in the wrong place. Please derive the runtime base with the same priority as core storage: `QWEN_RUNTIME_DIR` > `advanced.runtimeOutputDir` > `~/.qwen`.

_— gpt-5.5 via Qwen Code /review_

---
description: Chat session manager. /chat [-s|-l|-r|-d|-h] [name] [-y|--force]
---

# CRITICAL: First check {{args}}, then route

## Step 0: Immediate Validation (MUST execute FIRST)

**Check `{{args}}` right now, before doing anything else:**

1. Is `{{args}}` empty? → **Show Help immediately, STOP**
2. Is `{{args}}` only whitespace? → **Show Help immediately, STOP**
3. Does the first token look like a valid flag? (`-s`, `--save`, `-l`, `--list`, `-r`, `--resume`, `-d`, `--delete`, `-h`, `--help`)
   - **NO** → invalid flag/unrecognized → **Show Help immediately, STOP**
   - **YES** → Continue to Step 1

**⚠️ DO NOT skip this step. DO NOT proceed with any action until you verify `{{args}}`.**

---

## Step 1: Detect Environment

### Language

Run `node -e "console.log(Intl.DateTimeFormat().resolvedOptions().locale)"` to get system locale.
Use the language code (first 2 chars, e.g., "en", "zh", "ja") to determine response language.
If locale detection fails, match the language the user used in their prompt.

### OS Detection (ONLY for `-r`/`--resume`)

**Skip this step for other flags.** Only run when `-r` is detected.

> Run `node -e "console.log(process.platform)"`. Works across all shells.

Reviewer: **[Critical]** OS detection runs `node -e "console.log(process.platform)"` then routes by string. **WSL2 returns `linux`**, so the router routes to `chat-resume.md`'s Linux branch — `gnome-terminal -- bash -c "…"` or `xterm -e "…"`. Inside WSL these either fail (no X display) or pop up an X window the user can't easily reach, and `qwen --resume <id>` runs in that lost window.

WSL is a non-trivial slice of "Windows" qwen-code users. Repro path: "I `/chat -r foo`, nothing happens." No log to inspect (see also chat.md re: missing breadcrumbs). Hours of debug.

Suggested fix — when `process.platform === 'linux'`, additionally test for WSL:

```
If platform == 'linux':
  Read /proc/version. If it contains 'Microsoft' or 'WSL' (case-insensitive),
  treat as Windows for the resume branch — use Windows Terminal:
    wt.exe -d '<projectRoot>' -- qwen.exe --resume <id>
  (or fall back to: cmd.exe /c start cmd /k "cd /d <projectRoot> && qwen --resume <id>")
```

Document this branch in `chat-resume.md` Step 6 alongside the existing platform list.

_via Qwen Code /review_

- `win32` → Windows
- `linux` → Linux
- `darwin` → macOS

---

## Step 2: Parse and Route

Split `{{args}}` by whitespace. First token = flag. Remaining = raw_args.

| Flag              | Action                                    | Sub-Command File |
| ----------------- | ----------------------------------------- | ---------------- |
| `-s` / `--save`   | Go to Step 3                              | `chat-save.md`   |

> | `-l` / `--list`   | Read `chat-list.md` and execute its logic | `chat-list.md`   |

Reviewer: **[Suggestion]** The routing table instructs the LLM to "Read `chat-list.md` and execute its logic" / "Read corresponding sub-command file and execute". Per `packages/cli/src/services/command-factory.ts:125-150`, file commands return `submit_prompt` with the rendered markdown — there is **no built-in include/transclude mechanism** in the command processor. The pattern works only because the LLM voluntarily uses its file-reading tool when prompted. None of the 240 assertions in `test.mjs` exercise the actual routing — they all do `String.prototype.includes()` over the raw markdown.

Future-maintainer hazard: someone adding `/chat -e <name>` (export) will copy the row pattern, ship a feature that fails silently when an LLM (especially smaller / cheaper variants) decides to skip the file read. There's no test that catches it.

Suggested fix — pick one:

1. **Document the convention explicitly** in CHAT-DESIGN.md: "Routing relies on LLM tool-use, not runtime expansion. Verified via prompt behavior, not the test suite."
2. **Add at least one integration-style test** that boots the command processor, runs `/chat -l`, and asserts the LLM-facing prompt actually instructs reading `chat-list.md`.
3. **Inline the sub-commands** into chat.md if size budget allows — kills the routing-via-file-read pattern entirely.

_via Qwen Code /review_

| `-r` / `--resume` | Go to Step 3                              | `chat-resume.md` |
| `-d` / `--delete` | Go to Step 3                              | `chat-delete.md` |
| `-h` / `--help`   | **Show Help immediately, STOP**           | —                |

### Step 3: Validate name (for `-s`, `-r`, `-d`)

**For delete (`-d`):**

1. Parse raw_args to extract name: Filter out `-y` and `--force` flags first, the first remaining token is the name.
2. If name is missing, empty, or whitespace only → **Show Help immediately, STOP**
3. If extra non-flag tokens remain after the first name → **Show Help immediately, STOP**
4. If `-y` or `--force` was found → Set `forceDelete = true`

**For save/resume (`-s`, `-r`):**

> 1. Parse raw_args to extract name: Filter out any flags first, the first remaining token is the name.

Reviewer: **[Suggestion]** Step 3 says for save/resume: "Filter out any flags first, the first remaining token is the name." There's no definition of "any flag" here — only `-y`/`--force` are listed (and only for delete). `/chat -s my-name --garbage` is ambiguous: treat `--garbage` as a flag (filter, accept `my-name`, ignore typo)? Or extra non-flag token (reject)? Different LLM runs will answer differently, especially smaller ones.

Similarly `/chat -s my-name -y` — `-y` is documented as a delete-only option, but "filter out any flags" makes save/resume silently swallow it.

Suggested fix — make the rule symmetric with delete's explicit allowlist:

```
For save/resume: reject any token starting with `-` after the flag.
  Save and resume have no options.
  Error: "Unexpected token: <token>. /chat -s|-r takes only a single name."
For delete: allow only `-y` / `--force` after the name.
```

_via Qwen Code /review_

| **Valid name regex**  | `^[a-zA-Z0-9_.-]+$`                                                                                                                                                |
| **Max length**        | 128 characters                                                                                                                                                     |
| **Reserved names**    | `.`, `..`, `__proto__`, `constructor`, `prototype`                                                                                                                 |
| **Index path**        | `.qwen/chat-index.json` (project root)                                                                                                                             |
| **Index format**      | `{"name": "sessionId", ...}`                                                                                                                                       |
| **Session ID source** | Filename (no extension) of `.jsonl` in `<runtimeBase>/projects/<sanitizeCwd>/chats/`. runtimeBase priority: `$QWEN_RUNTIME_DIR` > `$QWEN_PROJECTS_DIR` > `~/.qwen` |

> | **Project dir**       | `sanitizeCwd(projectRoot)` replaces all non-alphanumeric characters with `-`. On Windows, also lowercase. E.g., `D:\code\qwen-code` → `d--code-qwen-code`          |

Reviewer: **[Critical]** The Runtime Base Resolution priority list documents `$QWEN_PROJECTS_DIR` as a fallback after `$QWEN_RUNTIME_DIR`. **`QWEN_PROJECTS_DIR` does not exist in the qwen-code codebase** — `grep -rn QWEN_PROJECTS_DIR packages/` returns zero hits. The same fictional precedence appears in `chat-save.md:13`, `chat-resume.md:17`, `chat-delete.md`, and the `chat-src/` source counterparts (5+ files in lockstep).

The **real** runtime base resolution is in `packages/core/src/config/storage.ts:103-119`:

1. `QWEN_RUNTIME_DIR` env var (matches the doc).
2. `setRuntimeBaseDir(path)` — driven by the `advanced.runtimeOutputDir` settings key.
3. `~/.qwen` default.

The `advanced.runtimeOutputDir` settings key is invisible to the LLM at command-execution time (it would have to read `~/.qwen/settings.json` — already rejected as a credential-leak risk in earlier rounds).

**Impact:** any user who configures `advanced.runtimeOutputDir` in settings will have sessions stored under the configured base, but `/chat -s|-l|-r|-d` will look under `~/.qwen` (because the LLM follows the docs and only checks two env vars) and report "No session found." These two flows silently disagree.

Suggested fix — replace `$QWEN_PROJECTS_DIR` references with reality:

```
Runtime Base Resolution (in priority order):
- $QWEN_RUNTIME_DIR (if set)
- ~/.qwen (default fallback)

Note: If the user has set `advanced.runtimeOutputDir` in settings.json, sessions
are under that path instead. /chat commands cannot read settings.json (credential
leak risk) and will not find those sessions. Document this limitation.
```

Fix in 5 files: `chat.md`, `chat-save.md`, `chat-resume.md` (production) and the chat-src counterparts.

_via Qwen Code /review_

## Help Text

**Show this when:**

- `{{args}}` is empty or whitespace only
- First token is NOT a valid flag
- Flag requires name but name is missing/empty
- User explicitly requests `-h` or `--help`

**Display this exact text and STOP all processing:**

```
Chat Session Manager

Usage: /chat <flag> [name] [-y|--force]

Flags:
  -s, --save <name>   Save current session with a name
  -l, --list          List all saved sessions
  -r, --resume <name> Resume a saved session
  -d, --delete <name> Delete a saved session from index
  -h, --help          Show this help

Options:
  -y, --force         Skip confirmation prompt (for -d)

Examples:
  /chat -s my-session
  /chat -l
  /chat -r my-session
  /chat -d my-session
  /chat -d my-session -y  # Delete without confirmation
```

> (End of file - 109 lines)

Reviewer: **[Suggestion]** This production command file ends with a copied file-reader artifact, `(End of file - 109 lines)`. Custom command Markdown is loaded as runtime instruction context, so this stray diagnostic text adds noise, can be interpreted as part of the command contract, and increases the production token budget. Please remove the trailing artifact line so the file ends after the help-text code block.

```suggestion

```

_— gpt-5.5 via Qwen Code /review_

   - **NO** �� invalid flag/unrecognized �� **Show Help immediately, STOP**
   - **YES** �� Continue to Step 1

**?? DO NOT skip this step. DO NOT proceed with any action until you verify `{{args}}`.**

---

## Step 1: Detect Environment
Read `./.qwen/settings.json` (project root).
If missing/not found → use `~/.qwen/settings.json` (global).

## Step 2: Detect Language
Read `./.qwen/settings.json` (project root).
Look for `general.language`.
If missing/not found → use `~/.qwen/settings.json` (global).
If still missing/not found → use `en`.
Supported languages: `zh`, `en`.

## Step 3: Parse Flags and Route
Split `{{args}}` by whitespace. First token = flag, rest = parameters.
Route based on flag:
| Flag | Action | File |
|------|--------|------|
| `-s` / `--save` `<name>` | Save session | `chat-save.md` |
| `-l` / `--list` | List sessions | `chat-list.md` |
| `-r` / `--resume` `<name>` | Resume session | `chat-resume.md` |
| `-d` / `--delete` `<name>` | Delete session | `chat-delete.md` |
| `-h` / `--help` | Show help | `chat-help.md` |
| (no flag) | New temporary session | `chat-new.md` |

## Step 4: Execute Command
Load the referenced `.md` file (from `.qwen/commands/`).

> Replace `{{name}}` with the provided name parameter (if any).

Reviewer: **[Critical]** The router assumes it can load another Markdown command file and replace `{{name}}`, but file-based commands are loaded as independent `submit_prompt` commands. The command processor substitutes the built-in `{{args}}` placeholder; it does not support a router dynamically loading another command file or templating a `{{name}}` placeholder.

This means `/chat -s foo`, `/chat -r foo`, and `/chat -d foo` can leave subcommands unexecuted or executed with literal `{{name}}`, so the unified `/chat` entry point is not implementable as written. Please either make each command independently invocable with `{{args}}`, or inline the subcommand logic into `chat.md`.

_— gpt-5.5 via Qwen Code /review_
